storage-proofs/post/src/fallback/vanilla.rs

```diff
@@ -507,82 +525,86 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
-                    .enumerate()
-                    .map(|(n, inclusion_proof)| -> Result<bool> {
-                        let challenge_index =
-                            (j * num_sectors_per_chunk + i) * pub_params.challenge_count + n;
-                        let challenged_leaf =
-                            generate_leaf_challenge_inner::<<Tree::Hasher as Hasher>::Domain>(
-                                challenge_hasher.clone(),
-                                pub_params,
-                                challenge_index as u64,
-                            );
-
-                        // validate all comm_r_lasts match
-                        if inclusion_proof.root() != comm_r_last {
-                            error!("inclusion proof root != comm_r_last: {:?}", sector_id);
-                            return Ok(false);
-                        }
-
-                        // validate the path length
-                        let expected_path_length = inclusion_proof
-                            .expected_len(pub_params.sector_size as usize / NODE_SIZE);
-
-                        if expected_path_length != inclusion_proof.path().len() {
-                            error!("wrong path length: {:?}", sector_id);
-                            return Ok(false);
-                        }
-
-                        if !inclusion_proof.validate(challenged_leaf as usize) {
-                            error!("invalid inclusion proof: {:?}", sector_id);
-                            return Ok(false);
-                        }
-                        Ok(true)
-                    })
-                    .collect::<Result<Vec<bool>>>()?;
-
-                let is_valid = is_valid_list.into_iter().all(|v| v);
-                if !is_valid {
-                    return Ok(false);
-                }
+                            // validate the path length
+                            let expected_path_length = inclusion_proof
+                                .expected_len(pub_params.sector_size as usize / NODE_SIZE);
+
+                            if expected_path_length != inclusion_proof.path().len() {
+                                error!("wrong path length: {:?}", sector_id);
+                                return Ok(false);
+                            }
+
+                            if !inclusion_proof.validate(challenged_leaf as usize) {
+                                error!("invalid inclusion proof: {:?}", sector_id);
+                                return Ok(false);
+                            }
+                            Ok(true)
+                        })
+                        .collect::<Result<Vec<bool>>>()?;
+
+                    Ok(is_valid_list.into_iter().all(|v| v))
+                })
+                .reduce(
+                    || Ok(true),
+                    |all_valid, is_valid| Ok(all_valid? && is_valid?),
+                )?;
+            if !is_valid {
+                return Ok(false);
```

Comment:
@dignifiedquire we also need continue the check and record the fault sectors here. I had lost the storage power cause some sectors failed in here, but i forgot the detail.